### ISSUE TITLE
Bad Oauth Request

### ISSUE THREAD

**Issue by cmwalolo** [500px/api-documentation#147]
https://api.500px.com/v1/oauth/access_token?oauth_callback=http://www.isc.ro
&oauth_consumer_key=u26X4av9ydNPd7kteT7bunfcdjHqVttYWIDOC1lA
&oauth_nonce=1618882
&oauth_signature_method=HMAC-SHA1
&oauth_timestamp=1433848367
&oauth_token=jBo5RRbNdQifFrDuzb4WZVFZVHNBc9j0mKkMhD1c
&oauth_verifier=XPuvcMNhqcugcPixct5e
&oauth_version=1.0
&oauth_signature=gcqpVyF7lO4qHWHQGzGgohQyWq4%253d

Invalid OAuth Request

Is there still something missing ?  

My Application is in my authorization lists, so I just need to get that one correctly.

**Comment by machx0r** [comment]
Are you trying to reuse an existing request token? If so you need to go through all three steps and then save the access token/secret until it's revoked. I'm not sure of the spec but 500px seemed to throw away the request token / verifier once it had been exchanged for an access token.

**Comment by cmwalolo** [comment]
Yah that is what I do. I use the code stated here : https://github.com/500px/api-documentation/issues/84
And added the oauth_verifier..

**Comment by machx0r** [comment]
I assume you're using the code at https://oauth.googlecode.com/svn/code/csharp/ correct? If so this is where I started also since it was less involved than the other options. I was not able to get any of the OAuth .NET libraries to work out of the box.

The **OAuthBase.cs** example had several problems but unfortunately I do not remember them all. 

Taking a quick peek again at the original code I see that the `GenerateSignatureBase()` will strip out the `oauth_verifier` parameter when it calls `GetQueryParameters()`. The `UrlEncode()` method is broken and there are several spots where a query parameter will either not be encoded OR double encoded.

Here's what I'm using for reference: http://pastebin.com/CwkudZ0G

Out of the box it's not going to work for you as I didn't include the `Interfaces` or `Contracts` but it might help you see where I made changes to the original **OAuthBase.cs** methods to get a working solution.

You'll also note that my example no longer contains the authentication methods as that was a hack and once I got a working access token and secret I moved on to actually consuming the API and planned to come back to the authorization stuff later.

Long story short I'm coming back to the authorization part now and will post a link to the complete project once I have something more complete :)

**Comment by cmwalolo** [comment]
Thank you ! Indeed I fixed a bit all the URLEncode, and found a double encoding in the original code.
But it still doesn't help ! still stuck at the third step...Two first steps looks OK as long my returned tokens are ok... 
I'm trying a bit more :=))

**Comment by machx0r** [comment]
There was a double encode bug in the file I uploaded earlier.

This:

`parameters.Add(new QueryParameter(OAuthParameter.Callback, UrlEncode(_callbackUrl)));`

Should be this:

`parameters.Add(new QueryParameter(OAuthParameter.Callback, _callbackUrl));`

I just implemented and verified the authentication steps in my solution so I'm uploading my code to GitHub in a minute.

**Comment by machx0r** [comment]
Here's the repo: https://github.com/machx0r/CCSWE-500px-API

Usage is straightforward. Windows Forms example:

    using System;
    using System.Diagnostics;
    using System.IO;
    using System.Net;
    using System.Text;
    using System.Threading;
    using System.Windows.Forms;
    using CCSWE.FiveHundredPx;

    namespace Test
    {
        public partial class Form1 : Form
        {
            public Form1()
            {
                InitializeComponent();

                _listener = new HttpListener();
                _listener.Prefixes.Add("http://localhost:1234/OAuth/");
                _listener.Start();

                var test2 = new Thread(Listen);
                test2.Start();

            }

            private static HttpListener _listener;

            private async void Listen()
            {
                while (true)
                {
                    var test = _listener.GetContext();
                    var response = test.Response;

                    var test2 = test.Request;
                    string responseString = "<HTML><BODY> Hello world!</BODY></HTML>";
                    byte[] buffer = Encoding.UTF8.GetBytes(responseString);
                    response.ContentLength64 = buffer.Length;
                    Stream output = response.OutputStream;
                    output.Write(buffer, 0, buffer.Length);
                    output.Close();

                    foreach (var key in test2.QueryString.AllKeys)
                    {
                        switch (key)
                        {
                            case OAuthParameter.Token:
                            {
                                _requestToken.Token = test2.QueryString[key];
                                break;
                            }
                            case OAuthParameter.Verifier:
                            {
                                _requestToken.Verifier = test2.QueryString[key];
                                break;
                            }
                        }
                    }

                    var accessToken = await _service.AccessToken(_requestToken);
                    _service.Token = accessToken;
                    var response3 = await _service.GetCurrentUser();
                }
            }

            private OAuthToken _requestToken;
            private FiveHundredPxService _service = new FiveHundredPxService("<CONSUMER KEY>", "<CONSUMER SECRET>", "http://localhost:1234/OAuth/");

            private async void button1_Click(object sender, EventArgs e)
            {
                _requestToken = await _service.RequestToken();
                webBrowser1.Navigate(_service.AuthorizeToken(_requestToken));
            }

            private void webBrowser1_Navigated(object sender, WebBrowserNavigatedEventArgs e)
            {
                Debug.WriteLine(e.Url);
            }
        }
    }

**Comment by cmwalolo** [comment]
That's cool :=) I try it right now ...

**Comment by cmwalolo** [comment]
And it's working... Wonderful... Many thanks Mach :=)
